### PR TITLE
[ONNXIFI] 4. Add class Tensor

### PR DESCRIPTION
**Pull request by apwojcik**
A wrapper class for ONNXIFI tensor descriptor.

Signed-off-by: Artur Wojcik <artur.wojcik@intel.com>

### CODE REVIEW

File: src/ngraph/frontend/onnxifi/tensor.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        };
+
+        class InputTensor final : public Tensor
+        {
+        public:
+            using Tensor::Tensor;
+            std::shared_ptr<runtime::Tensor> to_ng(runtime::Backend& backend) const final;
+        };
+
+        class OutputTensor final : public Tensor
```
Comment (postrational) [inline src/ngraph/frontend/onnxifi/tensor.hpp]: Please split the `tensor` file into separate files, one for each class.

File: src/ngraph/frontend/onnxifi/tensor.hpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            Tensor(Tensor&&) = default;
+            Tensor& operator=(Tensor&&) = default;
+
+            Tensor() = delete;
+            virtual ~Tensor() = default;
+
+            explicit Tensor(const ::onnxTensorDescriptorV1& tensor)
+                : m_tensor{&tensor}
+            {
+                if (tensor.tag != ONNXIFI_TAG_TENSOR_DESCRIPTOR_V1)
```
Comment (postrational) [inline src/ngraph/frontend/onnxifi/tensor.hpp]: Please keep your headers free of implementations.

File: src/ngraph/frontend/onnxifi/tensor.hpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+#include "ngraph/runtime/tensor.hpp"
+
+#include "exceptions.hpp"
+#include "span.hpp"
+
+namespace ngraph
+{
+    namespace onnxifi
+    {
+        /// \brief Wrapper to onnxTensorDescriptorV1 class
```
Comment (postrational) [inline src/ngraph/frontend/onnxifi/tensor.hpp]: `Wrapper to` -> `Wrapper for`

File: src/ngraph/frontend/onnxifi/tensor.hpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                        m_size *= value;
+                    }
+                }
+                if (tensor.buffer == 0)
+                {
+                    throw status::invalid_memory_location{};
+                }
+            }
+
+            virtual std::shared_ptr<runtime::Tensor> to_ng(runtime::Backend& backend) const = 0;
```
Comment (postrational) [inline src/ngraph/frontend/onnxifi/tensor.hpp]: Please add a line or two of comments for this function.

File: src/ngraph/frontend/onnxifi/tensor.hpp (hunk 1 above)
Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: Spliting into separate files will add complexity and increase compilation time due to extra preprocessor passes required to include those 5 lines + comple header comment + extra `#pragma one` guard stored in cache... `Tensor`, `OutputTensor` and `InputTensor` are intentionally placed in the single header file.

Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: Those 3 classes are logically and conceptually consistent and I prefer to keep it as is.

Comment (postrational) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: I would still rather see the classes split into separate files for easier navigation when editing code.

Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: done

File: src/ngraph/frontend/onnxifi/tensor.hpp (hunk 4 above)
Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: done

File: src/ngraph/frontend/onnxifi/tensor.hpp (hunk 3 above)
Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: done

File: src/ngraph/frontend/onnxifi/tensor.hpp (hunk 2 above)
Comment (apwojcik) [reply src/ngraph/frontend/onnxifi/tensor.hpp]: done

File: src/ngraph/frontend/onnxifi/input_tensor.hpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#include "ngraph/runtime/backend.hpp"
+#include "ngraph/runtime/tensor.hpp"
+
+#include "tensor.hpp"
+
+namespace ngraph
+{
+    namespace onnxifi
+    {
+        /// \brief Wrapper for ONNXIF input tensor
```
Comment (postrational) [inline src/ngraph/frontend/onnxifi/input_tensor.hpp]: Typo in `ONNXIF`

**Comment by apwojcik** [comment]
I removed distinction between `InputTensor` and `OutputTensor` in favour of just `Tensor` class realizing both `read` and `write` operations.

**Pull request merged**